### PR TITLE
Collections docs generation

### PR DESCRIPTION
**Pull request by abadger**
##### SUMMARY
Generate docs for plugins that are not hosted in ansible/ansible.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
docs
@acozine @samccann 

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by chouseknecht** [comment]
Is this waiting on this https://github.com/ansible/galaxy/issues/1992 ?

**Comment by abadger** [comment]
@chouseknecht yes-ish.... I think I can get this PR to a PoC stage without ansible/galaxy#1992 but I don't think it will be fast enough for production without some of the things in 1992.  There was also a lack of clarity around some of the other dependencies which is holding up full completion of this PR outside the control of 1992.

**Comment by abadger** [comment]
Right now, the following work:
* `make webdocs` will rebuild all documentation, including the collection docs
* `source ./hacking/env-setup ; ./hacking/build-ansible.py document-collection-plugins full` will build just the collection docs.

Need to do:
* Review by docs of the output to see if there's anything that needs changing
  * Text of the banner on collections pages
  * Text of the stub pages
  * URL/directory layout

Probably need to do:
* My understanding is that plugins are going to be split amongst hundreds of collections so we'll need to optimize this if we can:
  * The extra galaxy API from here https://github.com/ansible/galaxy/issues/1992 will help
  * Implementing incremental mode will help

Possibly need to do:
* Category pages for collections (catgeory = FQCN)

File: docs/docsite/collection-plugins.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# We also need an example of modules hosted in Automation Hub
```
Comment (gundalow) [inline docs/docsite/collection-plugins.yml:1]: I don't know if docs.ansible.com should ever point to docs on AH module/plugin docs.

File: docs/templates/moved_to_collections.rst.j2 [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+{%   set title = module %}
+{% endif %}
+
+@{ title }@
+@{ '+' * title|length }@
+
+This module has moved to a collection.  To install it::
+
+    ansible-galaxy collection install @{collection}@
+
```
Comment (gundalow) [inline docs/templates/moved_to_collections.rst.j2]: Possibly include a link collection docs here?

File: hacking/build_library/build_ansible/jinja2/tests.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+# if a module is added in a version of Ansible older than this, don't print the version added
+# information in the module documentation because everyone is assumed to be running something newer
+# than this already.
+TOO_OLD_TO_BE_NOTABLE = '2.4'
+
+test_list = partial(is_sequence, include_strings=False)
+
+
+def still_relevant(version, cutoff=TOO_OLD_TO_BE_NOTABLE):
```
Comment (gundalow) [inline hacking/build_library/build_ansible/jinja2/tests.py]: Nice :)

File: docs/docsite/collection-plugins.yml (hunk 1 above)
Comment (abadger) [reply docs/docsite/collection-plugins.yml:1]: <nid> yeah, if we only point at things in Galaxy, c then we won't need an automation hub hosted example

File: docs/templates/moved_to_collections.rst.j2 (hunk 2 above)
Comment (abadger) [reply docs/templates/moved_to_collections.rst.j2]: On galaxy or on docs.ansible.com?  The latter is at the bottom of the new template

**Comment by abadger** [comment]
We'll need this change: https://github.com/ansible/ansible/pull/66646 so that we can easily update the intersphinx cache files which link the collections docs site to the main ansible docs site.

**Comment by acozine** [comment]
We should add a line to the .gitignore for the local collections build, under https://github.com/ansible/ansible/pull/59761/files#diff-a084b794bc0759e7a6b77810e01874f2R32.

**Comment by ericsysmin** [comment]
How often will the docs be generated from the remote docstrings?

**Comment by mattclay** [comment]
@abadger What's the status of this? Does it have any relation to the work taking place in https://github.com/ansible-community/ansibulled?

**Comment by abadger** [comment]
@mattclay yep, after a request to redesign the docsite to document acd/ansible rather than ansible-base + a separate area for collections i had to go back and reevaluate how this would all work.

I'm in the process of shifting a lot of common code from hacking into ansibulled. Once that is done, this pr will need to be updated to remove a lot of code from hacking and instead running the ansibulled-docs program.

Not sure how tests of docsite build will work in the new system... I'm currently avoiding installing ansible at ansibulled install time but i am installing it at website build time because the set of plugins available and in need of documenting will depend on what is in the version of ansible-base associated with that ansible release.  From what we've talked about, i think that's even worse from the standpoint of the current ci test.  I think it's the way the website build has to work, though.

Talking with acozine and samccan i think maybe i could add a mode to the build where it only builds the docs that are in ansible/ansible.  Making ansible/ansible come from a directory passed on the command line instead of downloaded would be a further hack on top of that which might make a new test of website build in ansible's ci possible.

At the moment, though, I've been concentrating on getting the website built at all so i haven't spent time thinking through the new testing case

**Comment by abadger** [comment]
This bug is something of a blcoker: https://github.com/ansible/ansible/issues/69475   I can ping the collection author to fix the plugin in their collection but that's not a viable long term solution.  We'll need the ansible-doc tool to be more robust if we're going to use it for making collection documentation.

File: docs/docsite/collection-plugins.yml (hunk 1 above)
Comment (acozine) [reply docs/docsite/collection-plugins.yml:1]: agreed: docs.ansible.com should point to Galaxy only

**Comment by acozine** [comment]
Early versions of the new build are resource-intensive. We will come up with suggested minimum specs. 

If you see `OSError Errno 24 Too many open files` on macOS, increase your ulimit: `ulimit -n 1024`.

**Comment by abadger** [comment]
Rebased.   This is coming close to being ready.  I think most of the open bugs need changes in antsibull rather than in the code changes here.

One open question for docs: what to do with the pages in plugins/*.rst

* Perhaps the pages in plugins/*.rst should be deprecated
  altogether and their content integrated with other pages?
* If not, perhaps we want to rephrase and link into the collection
  documentation?
* Or perhaps we want to link to just the plugins which are present in
  collections/ansible/builtin?

**Comment by abadger** [comment]
I'm removing WIP because I think we're in the test, fix, test cycle now.

**Comment by samdoran** [comment]
!waffling ci_verified

**Comment by abadger** [comment]
Current time to run the build on the devel branch is about two minutes:
```
date>start  ; make webdocs ; date>end

Wed 08 Jul 2020 12:15:39 AM PDT
Wed 08 Jul 2020 12:17:34 AM PDT
```

This is because the devel branch only builds the docs for ansible-base.

**Comment by abadger** [comment]
In #70530 I found that the docs-build test doesn't test files which are marked `:orphan:` in the rst.  This applies to all of our module docs in ansible-2.9 and to all of our plugin (including modules) in ansible-2.10.  Luckily, with the vastly reduced number of modules, it isn't too big a difference to switch to make htmldocs which does build all of the documentation:

make singlehtmldocs:
Thu 09 Jul 2020 05:17:34 PM PDT
Thu 09 Jul 2020 05:18:50 PM PDT

make htmldocs
Thu 09 Jul 2020 05:20:26 PM PDT
Thu 09 Jul 2020 05:21:40 PM PDT

(ansible-test sanity --test docs-build will add a little overhead to both of those but it should be equal overhead)

Unfortunately, fixing the test shows that there's a lot of pre-existing problems with the module docs:
* https://gist.github.com/abadger/e2757f4354b2c79a4304f029ef1c5e39

(Separately, the first two errors are also not caught by ansible-test... I think I know why that is, though, and I can fix them).

@acozine, do we want to go ahead with this PR roughly how it is now and fix the test and the files which are broken later?  (Or I could fix the test case and add ignores entries for all the broken files).  (Some output will need to be ignored no matter what as those modules reference modules which are outside of ansible-base so the labels will be undefined in the ansible-base-only docs build which is what the new docs-build sanity test checks.)

**Comment by abadger** [comment]
After analyzing my gist, i think the majority of those errors will need to be ignored because they're external links.

There's about 12 which can be fixed.  (this is after applying https://github.com/ansible/ansible/pull/70530 )

**Pull request merged**